airbyte-integrations/connectors/source-harvest/source_harvest/streams.py

```diff
@@ -0,0 +1,296 @@
+        if stream_data.get("next_page"):
+            return {
+                "page": stream_data["next_page"],
+            }
+
+    def request_params(self, next_page_token: Mapping[str, Any] = None, **kwargs) -> MutableMapping[str, Any]:
+        params = super().request_params(next_page_token=next_page_token, **kwargs)
+        params["per_page"] = self.per_page
+        if next_page_token:
+            params.update(**next_page_token)
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        :return an iterable containing each record in the response
+        """
+        records = response.json().get(self.data_field, [])
+        yield from records
+
+
+class HarvestStreamIncrementalMixin(HttpStream, ABC):
+    cursor_field = "updated_at"
+
+    def __init__(self, updated_since: pendulum.datetime, **kwargs):
+        super().__init__(**kwargs)
+        self._updated_since = updated_since
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state object
+        and returning an updated state object.
+        """
+        latest_benchmark = latest_record[self.cursor_field]
+        if current_stream_state.get(self.cursor_field):
+            return {self.cursor_field: max(latest_benchmark, current_stream_state[self.cursor_field])}
+        return {self.cursor_field: latest_benchmark}
+
+    def request_params(self, stream_state: Mapping[str, Any], **kwargs) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state=stream_state, **kwargs)
+        updated_since = self._updated_since
+        if stream_state.get(self.cursor_field):
+            updated_since = stream_state[self.cursor_field]
+        params.update({"updated_since": updated_since})
+        return params
+
+
+class HarvestStreamWithPaginationSliced(HarvestStreamWithPagination):
+    @property
+    @abstractmethod
+    def parent_stream_name(self) -> str:
+        """
+        :return: parent stream class name
+        """
+
+    def stream_slices(self, sync_mode: SyncMode, cursor_field: List[str] = None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        parent_stream = type(self.parent_stream_name, (HarvestStreamWithPagination,), {})
+        items = parent_stream(authenticator=self.authenticator)
+        for item in items.read_records(sync_mode=sync_mode):
+            yield {"parent_id": item["id"]}
+
+
+class Contacts(HarvestStreamWithPagination, HarvestStreamIncrementalMixin):
+    pass
+
+
+class Clients(HarvestStreamWithPagination, HarvestStreamIncrementalMixin):
+    pass
+
+
+class Company(HarvestStream):
+    primary_key = None
+    data_field = None
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
```

Comment:
would not be need, see my comments above

Reply:
Company stream always has just one record, thus i found it more clear to just override `parse_response` method for that specific stream.

Reply:
I agree with the explicit approach, but we also need to think about the following things:

1. maintainability - our code should not require changes in the long term, just because we were strict
2. readability - the fact that the Company stream has only one record is a detail, that doesn't make difference from streams that have 0 or 1 records, it only adds `if/else` implementation where simple `if` would be enough.
3. Inheritance should cover real use cases, if we have too many entities for the sake of functional purity (btw, it wasn't met, as you have nonrelated methods in Paginated class) it is difficult to understand and follow. Things like pagination is an only small aspect of Streams work and it should not be a point of inheritance, but rather aggregation and delegation.